torch/nn/utils/rnn.py

```diff
@@ -109,10 +110,26 @@ def pack_padded_sequence(input, lengths, batch_first=False):
     """
     data, batch_sizes = PackPadded.apply(input, lengths, batch_first)
 
-    return PackedSequence(data, list(batch_sizes.data))
+    return PackedSequence(data, batch_sizes)
+
+
+def _symbolic_pack_padded_sequence(g, input, lengths, batch_first=False):
+    # There currently is no PackPadded operator in ONNX. We rely on an
+    # optimization pass to remove this later. It is an error if all
+    # PackPadded operators cannot be optimized out.
+    return g.op("PackPadded", input, lengths, outputs=2)
+
+
+def pack_padded_sequence(input, *args, **kwargs):
```

Comment:
I don't think you need to do this additional invocation - just apply `@symbolic_override_first_arg_based(_ _symbolic_pack_padded_sequence)` to the function below (see https://github.com/pytorch/pytorch/blob/5b43c22f73279c67084a2357a489420c705cb84f/torch/onnx/__init__.py#L381)

The reason why we were arguing about perf for the main RNN function was that the internal function `forward` was declared (and decorated) on each RNN() function invocation. In this case the decorator would run only once on module loading so the resulting code from decorator would be equivalent to what you wrote here.

Reply:
that would be great, but it doesn't work because importing torch.onnx at the top level is a circular dependency

Reply:
You've lost the public-facing docblock for `pack_padded_sequence` here. Move it from the internal function to this one.

Reply:
++

Reply:
argh, circular dependency is BS. We can extract those decorators from the onnx submodule and put them somewhere else. It's really ugly to proliferate these kind of hacks around. @ezyang - what's the regular practice for circular deps in pytorch? Was there some design of which submodules can depend on others?

Reply:
I'd rather fix this as a separate issue; this diff doesn't lock us into doing it this way if someone finds a better way down the line

Reply:
There isn't a circular dependency discipline at the moment; but I would definitely not oppose moving these combinators to another module. Our current practice is, however, to push onnx imports inside functions precisely to avoid this problem.